### PR TITLE
Init translation of PWA caching

### PR DESCRIPTION
**Pull request by hanyujie2002**
### Description

Init translation of PWA caching

### Motivation

Ditto

### Additional details

[original link](https://github.com/mdn/content/blob/main/files/en-us/web/progressive_web_apps/guides/caching/index.md?plain=1)

In line 134 of original content(line 133 in translated content), there is a term "stale while revalidate", I don't have any good idea about translating it 😢

### CODE REVIEW

File: files/zh-cn/web/progressive_web_apps/guides/caching/index.md [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+缓存策略在离线操作、响应速度和新鲜度之间寻求平衡。不同的资源在此方面有不同的要求：例如，应用的基本 UI 可能相对稳定，而显示产品列表时获取最新数据则可能至关重要。这意味着 PWA 通常会对不同的资源采用不同的策略，单个 PWA 可能会使用此处描述的所有策略。
+
+### 缓存优先
+
+在这个策略中，我们会预缓存一些资源，然后仅对这些资源实现“缓存优先”策略。也就是：
+
+- 对于预缓存的资源，我们将：
+  - 在缓存中查找该资源，如果找到则返回该资源。
+  - 否则，转到网络。如果网络请求成功，则缓存该资源以备下次使用。
```
Comment (jasonren0403) [inline files/zh-cn/web/progressive_web_apps/guides/caching/index.md:73]:
```suggestion
  - 否则，进行一次网络请求。如果网络请求成功，则缓存该资源以备下次使用。
```

File: files/zh-cn/web/progressive_web_apps/guides/caching/index.md [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+- 对于所有其他资源，我们将始终转到网络。
```
Comment (jasonren0403) [inline files/zh-cn/web/progressive_web_apps/guides/caching/index.md:74]:
```suggestion
- 对于所有其他资源，我们将始终进行网络请求。
```

File: files/zh-cn/web/progressive_web_apps/guides/caching/index.md [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+预缓存适用于 PWA 确定需要的、在此版本的应用中不会改变的资源，以及需要尽快获取的资源。这包括，例如，应用的基本用户界面。如果这些界面预缓存了，那么应用的 UI 就可以在启动时不需要任何网络请求地渲染。
```
Comment (jasonren0403) [inline files/zh-cn/web/progressive_web_apps/guides/caching/index.md:76]:
```suggestion
预缓存适用于 PWA 确定需要的、在此版本的应用中不会改变的资源，以及需要尽快获取的资源。例如，这包括应用程序的基本用户界面。如果这些界面预缓存了，那么应用的 UI 就可以在启动时不需要任何网络请求地渲染。
```

File: files/zh-cn/web/progressive_web_apps/guides/caching/index.md [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+}
+
+self.addEventListener("fetch", (event) => {
+  if (precachedResources.includes(url.pathname)) {
+    event.respondWith(cacheFirst(event.request));
+  }
+});
+```
+
+我们通过调用事件的 {{domxref("FetchEvent.respondWith()", "respondWith()")}} 方法来返回资源。如果我们没有为某个请求调用 `respondWith()`，那么该请求将像 service worker 没有拦截它一样发送到网络。因此，如果一个请求没有预缓存，它就直接去网络。
```
Comment (jasonren0403) [inline files/zh-cn/web/progressive_web_apps/guides/caching/index.md:123]:
```suggestion
我们通过调用事件的 {{domxref("FetchEvent.respondWith()", "respondWith()")}} 方法来返回资源。如果我们没有为某个请求调用 `respondWith()`，那么该请求将像 service worker 没有拦截它一样发送到网络。因此，如果一个请求没有预缓存，它就直接从网络获取。
```

File: files/zh-cn/web/progressive_web_apps/guides/caching/index.md [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+
+当我们将 `networkResponse` 添加到缓存时，我们必须克隆响应并将副本添加到缓存中，同时返回原始响应。这是因为 `Response` 对象是可以流传输的，所以只能读取一次。
+
+你可能会疑惑为什么对预缓存的资源会回退到网络。如果它们已预缓存，难道我们就不能确定它们一定在缓存中吗？原因是缓存可能会被浏览器或用户清除。虽然这不太可能发生，但如果不能回退到网络，会导致 PWA 不可用。参见[删除缓存数据](#删除缓存数据)。
```
Comment (jasonren0403) [inline files/zh-cn/web/progressive_web_apps/guides/caching/index.md:127]:
```suggestion
你可能会疑惑为什么对预缓存的资源会回退到网络请求。如果它们已预缓存，难道我们就不能确定它们一定在缓存中吗？原因是缓存可能会被浏览器或用户清除。虽然这不太可能发生，但如果不能回退到网络请求，会导致 PWA 不可用。参见[删除缓存数据](#删除缓存数据)。
```

File: files/zh-cn/web/progressive_web_apps/guides/caching/index.md [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+### 带缓存刷新的数据缓存优先
+
+“缓存优先”的缺点是一旦响应进入缓存，它就永远不会刷新，直到安装新的 service worker 版本。
+
+“带缓存刷新的数据缓存优先”策略，也称为“重新验证过时数据（stale while revalidate）”，与“缓存优先”类似，不同之处在于即使缓存命中后，我们也总是将请求发送到网络，并使用响应来刷新缓存。这意味着我们可以得到“缓存优先”的响应速度，但可以获取相当新鲜的响应（只要请求频繁足够）。
```
Comment (jasonren0403) [inline files/zh-cn/web/progressive_web_apps/guides/caching/index.md:133]: HTTP 中有[类似的stale-while-revalidate指令](https://datatracker.ietf.org/doc/html/rfc5861#section-3)，可以参考下，或者是[这篇专栏](https://zhuanlan.zhihu.com/p/64694485)

File: files/zh-cn/web/progressive_web_apps/guides/caching/index.md [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+    event.respondWith(cacheFirstWithRefresh(event.request));
+  }
+});
+```
+
+注意，我们异步更新缓存（在 `then()` 处理程序中），所以应用不必等待网络响应被接收就可以使用缓存的响应。
+
+### 网络优先
+
+我们将看到的最后一个策略是“网络优先”，它与缓存优先的逻辑相反：我们先尝试从网络获取资源。如果网络请求成功，我们返回响应并更新缓存。如果失败，我们尝试缓存。
```
Comment (jasonren0403) [inline files/zh-cn/web/progressive_web_apps/guides/caching/index.md:168]:
```suggestion
我们要了解的最后一个策略是“网络优先”，它与缓存优先的逻辑相反：我们先尝试从网络获取资源。如果网络请求成功，我们返回响应并更新缓存。如果失败，我们尝试获取缓存。
```

File: files/zh-cn/web/progressive_web_apps/guides/caching/index.md [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+这对于获取最新响应非常重要，但缓存的资源比没有好的请求很有用。一个消息应用的最近消息列表可能属于此类别。
+
+在下面的示例中，我们对获取位于应用 "inbox" 路径下的所有资源的请求使用“网络优先”。
```
Comment (jasonren0403) [inline files/zh-cn/web/progressive_web_apps/guides/caching/index.md:172]:
```suggestion
在下面的示例中，我们对获取位于应用“inbox”路径下的所有资源的请求使用“网络优先”策略。
```

File: files/zh-cn/web/progressive_web_apps/guides/caching/index.md [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+    event.respondWith(networkFirst(event.request));
+  }
+});
+```
+
+仍然存在一些请求，没有响应比可能过时的响应更好，且仅“网络优先”策略才合适。例如，如果一个应用正在显示可用产品列表，列表过时对用户来说将是令人沮丧的。
+
+## 删除缓存数据
+
+缓存具有有限的存储空间，如果超出限制，浏览器可能会逐出应用缓存的数据。具体的限制和行为因浏览器而异：参见[存储配额和驱逐标准](/zh-CN/docs/Web/API/Storage_API/Storage_quotas_and_eviction_criteria)了解详细信息。在实际中，驱逐缓存数据是一个非常罕见的事件。用户也可以随时清除应用的缓存。
```
Comment (jasonren0403) [inline files/zh-cn/web/progressive_web_apps/guides/caching/index.md:201]:
```suggestion
缓存具有有限的存储空间，如果超出限制，浏览器可能会驱逐应用缓存的数据。具体的限制和行为因浏览器而异：参见[存储配额和驱逐标准](/zh-CN/docs/Web/API/Storage_API/Storage_quotas_and_eviction_criteria)了解详细信息。实际上，驱逐缓存数据的情况非常罕见。用户也可以随时清除应用程序的缓存。
```

File: files/zh-cn/web/progressive_web_apps/guides/caching/index.md [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+
+PWA 应该在 service worker 的 {{domxref("ServiceWorkerGlobalScope.activate_event", "activate")}} 事件中清除旧版本的任何缓存：当此事件触发时，service worker 可以确定没有之前版本的 service worker 正在运行，因此不再需要旧的缓存数据。
+
+## 参见
+
+- [Service Worker API](/zh-CN/docs/Web/API/Service_Worker_API)
+- [Fetch API](/zh-CN/docs/Web/API/Fetch_API)
+- [存储配额和驱逐标准](/zh-CN/docs/Web/API/Storage_API/Storage_quotas_and_eviction_criteria)
+- [service worker 缓存策略](https://developer.chrome.com/docs/workbox/caching-strategies-overview/) on developer.chrome.com (2021)
+- [离线应用程序开发指南](https://web.dev/offline-cookbook/) on web.dev (2020)
```
Comment (jasonren0403) [inline files/zh-cn/web/progressive_web_apps/guides/caching/index.md:211]:
```suggestion
- developer.chrome.com 上的 [service worker 缓存策略](https://developer.chrome.com/docs/workbox/caching-strategies-overview/)（2021）
- web.dev 上的[离线应用程序开发指南](https://web.dev/offline-cookbook/)（2020）
```

**Review by jasonren0403 — COMMENTED** [review]
一些建议

**Comment by hanyujie2002** [comment]
Hi @jasonren0403, do you have any recommendation on the translation of `stale-while-revalidate`?

**Comment by jasonren0403** [comment]
> Hi @jasonren0403, I translate `stale while revalidate` into “用旧并换新”，do you have any recommendation on the translation of `stale-while-revalidate`?

In [HTTP Cache-control document](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control), `stale-while-revalidate` means that 'after the response becomes stale(after max-age), the client can still use the response for some time, but they should do the revalidation(probably by making a network request) on the same time to keep the response fresh. The revalidation happens async(or background) so that the user don't have to wait for the revalidated response.' 

Therefore I would suggest '异步缓存更新' or '后台缓存刷新'.

**Comment by yin1999** [comment]
How about "异步缓存更新"

**Comment by hanyujie2002** [comment]
In my opinion, the `stale-while-revalidate` are a strategy that not only related to cache which is reminded above, `stale-while-revalidate` could be also used for http request [HTTP Cache-control document](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control). What's more, there is no `cache` in the term `stale-while-revalidate`, so I think we don't need to include "缓存" in the translation, maybe there are better Chinese translation than "用旧并换新", but I think this translation is good enough to be understood by the Chinese readers while remaining most of the meaning of `stale while revalidate` @jasonren0403

**Comment by hanyujie2002** [comment]
Please merge ( •̀ ω •́ )✧

**Review by jasonren0403 — APPROVED** [review]
Thank you @hanyujie2002 !

**Pull request merged**